### ISSUE TITLE
Unable to use abstract types

### ISSUE THREAD

**Issue by berezovskyi** [eclipse/lyo.core#116]
I am trying to take advantage of the patch #29 by @neormx.

I tried replacing the code I had for naive unmarshalling of multiple classes (ie try this, if you get an exception try that and so on) with the code that simply unmarshalls a superclass (code in Kotlin):

```kotlin
//            val action: IExtendedResource = OslcHelpers.navTry(plan, step.action,
//               DropBelt::class.java, MoveToWp::class.java,
//                PickShelf::class.java)
            val action: IExtendedResource = OslcHelpers.navTry(plan, step.action,
                ExecutableAction::class.java)
```

The code fails. I tried the following:

- replace ExecutableAction with IExecutableAction (both generated with Lyo). Fail - because IExecutableAction generated by Lyo is an interface without any annotations and does not extend IResource. There is no way to put that one in JMH.
- make ExecutableAction abstract. I recall that Ricardo wanted to make it work only if the class you passed was explicitly abstract or an interface to avoid breaking any code. No luck. The problem here seems to be that ExecutableAction is annotated with an RDF type and JMH looks for it exactly; without adding RDFS reasoning it's impossible to do so (plus LyoD should them generate an RDFS vocab that can be used by Jena reasoner). `model.listSubjectsWithProperty` returns 0 resources.
- remove OSLC annotations from the generated ExecutableAction class. The Pet class from Ricardo's unit tests did not have any annotations, so I have removed mine too. ~This actually moved the debugging forward, now I get 1 resource in the `resourceList`! At this stage, however, a completely wrong package got scanned in the `createObjectResultList`. For some reason, the `beanClass` was set to the class name which holds properties of abstract types.~ False hope, that was accidental trigger of the debugger with an unrelated JMH call.

The classes discussed here are located in https://github.com/EricssonResearch/scott-eu/tree/master/lyo-services/domain-pddl/src/main/java/eu/scott/warehouse/domains/scott

Thanks in advance!

**Comment by neormx** [comment]
Hello @berezovskyi, not sure whether I understood the problem. I'm supposing:

1. `OslcHelpers.navTry` method is calling `JMH.unmarshal` internally.
2. Parameter `plan` is a Jena model.
3. You was expecting the most concrete class among `DropBelt`, `MoveToWp` and `PickShelf` was automatically discovered from the model by passing the common super class of them: `ExecutableAction`.

If all assumptions above are correct, JMH was not enhanced by #29 to do this. #29 enhanced JMH to support abstract types but only for the properties (and nested properties) of the main class to instantiate which needs to be passed always as the second argument to `JMH.fromJenaModel` method.

In other words, it's still up to the programmer to decide which class will be the main container for the RDF deserialization:

`ResouceShape[] rs = (ResouceShape[]) JenaModelHelper.fromJenaModel(model, ResouceShape.class);`

If my assumptions are wrong, could you please provide me a more precise scenario I could test to replicate the issue?

**Comment by berezovskyi** [comment]
> You was expecting the most concrete class among DropBelt, MoveToWp and PickShelf was automatically discovered from the model by passing the common super class of them: ExecutableAction.

Yes, @neormx, that was exactly my use-case. Sad to hear it's not supported. I will check what extra support needs to be added to JMH to cover it and consult with you.

**Comment by neormx** [comment]
I've created PR #117 to address this. Please review it and let's discuss any concern.
